src/coreclr/jit/objectalloc.cpp

```diff
@@ -1335,3 +1483,1734 @@ void ObjectAllocator::RewriteUses()
         }
     }
 }
+
+//------------------------------------------------------------------------------
+// AnalyzeIfCloningCanPreventEscape: see if by cloning we can ensure an object
+//    does not escape.
+//
+// Arguments:
+//   bitVecTraits                       - Bit vector traits
+//   escapingNodes               [in]   - current set of escaping nodes
+//   escapingNodesToProcess  [in/out]   - set of newly escaping nodes
+//
+// Returns:
+//   true, if there are any newly escaping nodes
+//
+// Notes:
+//   During our analysis we have may have noted conditionally escaping objects
+//   and var references and connected them to a pseduolocal, along with information
+//   about how we could clone blocks to ensure that the object could be stack allocated.
+//
+//   The current assumption is that these nodes do not escape, but to ensure
+//   that we must be able to clone the code and remove the potential for escape
+//
+//   So, we  verify for each case that we can clone; if not, mark we the pseudolocal
+//   as escaping. If any pseudlocal now escapes, we return true so that the main
+//   analysis can update its closure.
+//
+//   We may choose not to clone a candiate for several reasons:
+//   * too much EH already in the method, or some other reason cloning is infeasible
+//   * two different candidates have overlapping clone regions
+//   * the cost/benefit analysis does not look favorable
+//
+bool ObjectAllocator::AnalyzeIfCloningCanPreventEscape(BitVecTraits* bitVecTraits,
+                                                       BitVec&       escapingNodes,
+                                                       BitVec&       escapingNodesToProcess)
+{
+    bool newEscapes = false;
+
+    for (unsigned p = 0; p < m_numPseudoLocals; p++)
+    {
+        unsigned const pseudoLocal = p + comp->lvaCount;
```

Comment:
Does the base analysis actually need to allocate an index for all locals in the first place? Shouldn't it be sufficient to allocate indices for only `TYP_REF, TYP_BYREF and TYP_STRUCT` locals?

Reply:
Yeah, it is using bigger BVs than necessary. We would need to create some convenient dense numbering scheme. I can revisit this when I look at field-sensitive approaches perhaps.

Reply:
We could probably just reuse `LclVarDsc::lvVarIndex` for the dense numbering scheme. Doing this later sounds good to me.

It also seems like there should be some kind of limit on the number of tracked locals here, since this analysis is inherently super-linear.